cpp/sopt/gradient_utils.h

```diff
@@ -0,0 +1,34 @@
+#ifndef SOPT_GRADIENT_UTILS_H
+#define SOPT_GRADIENT_UTILS_H
+
+#include "sopt/types.h"
+#include "sopt/linear_transform.h"
+#include <memory>
+
+namespace sopt {
+
+//! A helper object that can hold the measurement operator
+//! and the target vector for a given iteration state
+template <typename T>
+class IterationState {
+  public:
+  IterationState() = delete;
+
+  IterationState(const T& target_vis,
+                 std::shared_ptr<sopt::LinearTransform<T>> phi)
+                 : _target(target_vis) {
```

Comment:
```suggestion
  IterationState(const T& target,
                 std::shared_ptr<sopt::LinearTransform<T>> phi)
                 : _target(target) {
```

can we change the name from `target_vis` so that we don't suggest that the target is supposed to be a visibility? In principle it could be anything.